### PR TITLE
enable equal_nan for complex values in isclose

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63841
* #63572
* __->__ #63571

Differential Revision: [D30560127](https://our.internmc.facebook.com/intern/diff/D30560127)

### CODE REVIEW

File: aten/src/ATen/native/TensorCompare.cpp [hunk 1]
```diff
@@ -121,8 +119,8 @@ Tensor isclose(const Tensor& self, const Tensor& other, double rtol, double atol
 
   // Computes equality closeness
   Tensor close = self == other;
-  if (equal_nan && self.is_floating_point()) {
-      close.__ior__((self != self).__iand__(other != other));
+  if (equal_nan && (self.is_floating_point() || self.is_complex())) {
+      close.__ior__(self.isnan().__iand__(other.isnan()));
```
Comment (pmeier) [inline aten/src/ATen/native/TensorCompare.cpp:123]: Although it is functionally equal, `self.isnan()` is a lot more legible than `self != self`

File: test/test_testing.py [hunk 2]
```diff
@@ -416,13 +414,20 @@ def test_isclose_comparetensors_complex(self, device, dtype):
...
-        with self.assertRaises(RuntimeError):
-            self._isclose_helper(tests, device, dtype, True)
-
+        tests = (
+            (complex(1, 1), complex(1, float('nan')), False),
+            (complex(1, 1), complex(float('nan'), 1), False),
+            (complex(float('nan'), 1), complex(float('nan'), 1), True),
+            (complex(float('nan'), 1), complex(1, float('nan')), False),
+            (complex(float('nan'), float('nan')), complex(float('nan'), float('nan')), True),
+        )
```
Comment (pmeier) [inline test/test_testing.py:430]: The test cases will be merged into the ones above later in the stack (#63572), when `TestCase.assertEqual` stops treating complex components separately.

File: test/test_testing.py [hunk 3]
```diff
@@ -449,6 +454,19 @@ def test_isclose_equality_shortcut(self):
 
         self.assertFalse(torch.isclose(a, b, rtol=0, atol=0))
 
+    @dtypes(torch.float16, torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_isclose_nan_equality_shortcut(self, device, dtype):
+        if dtype.is_floating_point:
+            a = b = torch.nan
+        else:
+            a = complex(torch.nan, 0)
+            b = complex(0, torch.nan)
```
Comment (pmeier) [inline test/test_testing.py:463]: We can also split this into two tests for float and complex.

File: aten/src/ATen/native/TensorCompare.cpp (hunk 1 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorCompare.cpp:123]: I guess so. The compiler can probably optimize away the extra function call.

**Review by mruberry — APPROVED** [review]
Cool! And it looks like the docs don't need to be updated, either.

**Pull request closed without merging**